pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,318 @@
+                        obs_new = [qml.PauliZ(wires=i) for i in m.wires]
+
+                obs_new.append(qml.PauliY(wires=aux_wire))
+                obs_new = qml.operation.Tensor(*obs_new)
+
+                if isinstance(m, qml.measurements.ExpectationMP):
+                    measurements.append(qml.expval(op=obs_new))
+                else:
+                    measurements.append(qml.probs(op=obs_new))
+
+            new_tape = qml.tape.QuantumTape(ops=ops, measurements=measurements)
+
+            def stop_at(obj):
+                return ~qml.operation.is_measurement(obj)
+
+            # Expand measurements only
+            new_tape = qml.tape.tape.expand_tape(
+                new_tape, stop_at=stop_at, expand_measurements=True
+            )
+            num_tape += 1
+
+            g_tapes.append(new_tape)
+
+        gradient_data.append(num_tape)
+
+    def processing_fn(results):  # pylint: disable=too-many-branches
+        # TODO interface compatibility
+        multi_measurements = len(tape.measurements) > 1
+        multi_params = len(tape.trainable_params) > 1
+
+        grads = []
+        final_res = []
+        for coeff, res in zip(coeffs, results):
+            if isinstance(res, tuple):
+                final_res.append([2 * coeff * r for r in res])
+            else:
+                final_res.append(2 * coeff * res)
+
+        # Post process for probs
+        if measurements_probs:
+            projector = np.array([1, -1])
+            for idx, res in enumerate(final_res):
+                if multi_measurements:
+                    for prob_idx in measurements_probs:
+                        prob = tape.measurements[prob_idx]
+                        wires = prob.wires
+                        num_wires = len(wires)
+                        res_reshaped = qml.math.reshape(res[prob_idx], (2**num_wires, 2))
+                        final_res[idx][prob_idx] = qml.math.tensordot(
+                            res_reshaped, projector, axes=[[1], [0]]
+                        )
+                else:
+                    prob_idx = measurements_probs[0]
+                    prob = tape.measurements[prob_idx]
+                    wires = prob.wires
+                    num_wires = len(wires)
+                    res = qml.math.reshape(res, (2**num_wires, 2))
+                    final_res[idx] = qml.math.tensordot(res, projector, axes=[[1], [0]])
+
+        for idx, num_tape in enumerate(gradient_data):
+            if num_tape is None:
+                grads.append(qml.math.zeros(()))
+            elif num_tape == 1:
+                grads.append(final_res[idx])
+            else:
+                grads.append(sum(final_res[idx : idx + num_tape]))
+
+        if not multi_measurements and not multi_params:
+            return np.array(grads[0])
+
+        if (multi_params and not multi_measurements) or (not multi_params and multi_measurements):
+            grads_tuple = tuple(grads)
+            return grads_tuple
```

Comment:
you already know at least one is true :)
```suggestion
        if not (multi_params and multi_measurements):
            return tuple(grads)
```